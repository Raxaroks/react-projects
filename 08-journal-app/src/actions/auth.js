import Swal from 'sweetalert2';
import { types } from "../types/types";
import { googleAuthProvider, firebase } from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";

export const startLoginWithEmailPassword = ( email, password ) => {
    return (dispatch) => {
        dispatch( startLoading() );
        firebase.auth().signInWithEmailAndPassword( email, password )
            .then( ({ user }) => {
                dispatch( login(user.uid, user.displayName) );
                dispatch( finishLoading() );
            } )
            .catch( error => {
                // console.log(error);
                dispatch( finishLoading() );
                Swal.fire( 'Error', error.message, 'error' );
            } );
    }
} 

export const startRegisterWithEmailNamePassword = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword( email, password )
            .then( async ({ user }) => {
                // console.log(user);
                await user.updateProfile({ displayName: name })
                dispatch( login( user.uid, user.displayName ) );
            } )
            .catch( error => {
                // console.log(error);
                Swal.fire( 'Error', error.message, 'error' );
            } );
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ( { user } ) => {

                dispatch(
                    login( user.uid, user.displayName )
                )

            })
            .catch( error => {
                console.log(error);
            } );
    }
}

export const login = ( uid, displayName ) => ({
    type: types.login,
    payload: {
        uid, 
        displayName
    }
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch( logout() );
    }
}

export const logout = () => ({
    type: types.logout
})