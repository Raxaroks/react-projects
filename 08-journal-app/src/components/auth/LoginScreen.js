import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startGoogleLogin, startLoginWithEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';


const LoginScreen = () => {

    const dispatch = useDispatch();
    const { loading } = useSelector( state => state.ui );

    const [ formValues, handleInputChange ] = useForm({
        email: 'devops_user@test.com',
        password: 'password123'
    });

    const { email, password } = formValues;

    const handleLogin = (evt) => {
        evt.preventDefault();

        if ( isFormValid() ) {
            // console.log('El formulario es correcto.');
            dispatch( startLoginWithEmailPassword( email, password ) );    
        }
    }

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    }

    const isFormValid = () => {
        if ( !validator.isEmail( email ) ) {
            // console.log('El email no es v치lido');
            dispatch( setError('El email no es v치lido') );
            return false;
        }

        else if ( password < 5 ) {
            // console.log('La contrase침a no es segura o no coincide.');
            dispatch( setError('La contrase침a no es segura o no coincide.') );
            return false;
        }

        dispatch( removeError() );
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form onSubmit={ handleLogin }>
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <button
                    className="btn btn-primary btn-block"
                    type="submit"
                    disabled={ loading }
                >
                    Login
                </button> 

                <div className="auth__social-networks">
                    <p><strong>Login con redes sociales</strong></p>
                    <div className="google-btn" onClick={ handleGoogleLogin }>
                        <div className="google-icon-wrapper">
                            <img 
                                className="google-icon"
                                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                                alt="Logo de Google"
                            />
                        </div>
                        <p className="btn-text">
                            <b>Login con Google</b>
                        </p>
                    </div>
                </div>

                <Link className="link" to="/auth/register">
                    Crea una nueva cuenta
                </Link>
            </form>
        </>
    );
}

export default LoginScreen;
