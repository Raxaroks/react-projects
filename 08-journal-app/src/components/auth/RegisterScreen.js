import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailNamePassword } from '../../actions/auth';


const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { errorMsg } = useSelector( state => state.ui );
    
    const [ formValues, handleInputChange ] = useForm({
        name: 'Andrés Ornelas',
        email: 'devops_user@test.com',
        password: 'password123',
        password2: 'password123'
    });

    const { name, email, password, password2 } = formValues; 

    const handleRegister = (evt) => {
        evt.preventDefault();
        // console.log( name, email, password, password2 );

        if ( isFormValid() ) {
            // console.log('El formulario está vergas.');
            dispatch( startRegisterWithEmailNamePassword( email, password, name ) );
        }

    } 

    const isFormValid = () => {

        if ( name.trim().length === 0 ) {
            // console.log('El nombre no está completo.');            
            dispatch( setError('El nombre no está completo.') );
            return false; 
        }

        else if ( !validator.isEmail( email ) ) {
            // console.log('El email no es válido');
            dispatch( setError('El email no es válido') );
            return false;
        }

        else if ( password !== password2 || password.length < 5 ) {
            // console.log('La contraseña no es segura o no coincide.');
            dispatch( setError('La contraseña no es segura o no coincide.') );
            return false;
        }

        dispatch( removeError() );
        return true;

    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form onSubmit={ handleRegister }>

                {
                    errorMsg &&
                    (
                        <div className="auth__alert-error">
                            { errorMsg }
                        </div>
                    )
                    
                }

                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    value={ name }
                    className="auth__input"
                    autoComplete="off"
                    onChange={ handleInputChange }
                />
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    value={ email }
                    className="auth__input"
                    autoComplete="off"
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    value={ password }
                    className="auth__input"
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    value={ password2 }
                    className="auth__input"
                    onChange={ handleInputChange }
                />
                <button
                    className="btn btn-primary btn-block mb-5"
                    type="submit"
                    // disabled={ true }
                >
                    Regístrate
                </button> 

                <Link className="link mt-1" to="/auth/login">
                    ¿Ya estás registrado? ¡Inicia sesión!
                </Link>
            </form>
        </>
    );
}

export default RegisterScreen;