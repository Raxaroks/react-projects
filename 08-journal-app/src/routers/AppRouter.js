import React, { useEffect, useState } from 'react';
import { BrowserRouter as Router, Redirect, Switch } from 'react-router-dom';
import JournalScreen from '../components/journal/JournalScreen';
import AuthRouter from './AuthRouter';
import { firebase } from '../firebase/firebase-config';
import { useDispatch } from 'react-redux';
import { login } from '../actions/auth';
import PrivateRoute from './PrivateRoute';
import PublicRoute from './PublicRoute';

const AppRouter = () => {

    const dispatch = useDispatch();

    const [checking, setChecking] = useState(true);
    const [isLoggedIn, setIsLoggedIn] = useState(false);

    useEffect( () => {
        firebase.auth().onAuthStateChanged( (user) => {
            // console.log(user);
            if ( user?.uid ) {
                dispatch( login( user.uid, user.displayName ) );
                setIsLoggedIn(true);
            }
            
            else {
                setIsLoggedIn(false);
            }

            setChecking(false);
        } );
    }, [ dispatch, setChecking, setIsLoggedIn ] ); // s√≥lo se ejecuta una vez

    if (checking) {
        return (
            <h1>Espere...</h1>
        )
    }

    return (
        <Router>
            <div>
                <Switch>
                    
                    <PublicRoute
                        path="/auth"
                        component={ AuthRouter }
                        isAuthenticated={ isLoggedIn }
                    />

                    <PrivateRoute
                        exact
                        path="/"
                        component={ JournalScreen }
                        isAuthenticated={ isLoggedIn }
                    />

                    <Redirect to="/auth/login" />

                </Switch>
            </div>
        </Router>
    );
}

export default AppRouter;
