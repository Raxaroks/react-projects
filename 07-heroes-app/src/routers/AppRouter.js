import React, { useContext } from 'react';
import {
  BrowserRouter as Router,
  Switch,
} from "react-router-dom";
import { AuthContext } from '../auth/AuthContext';
import PrivateRoute from './PrivateRoute';
import PublicRoute from './PublicRoute';
import LoginScreen from '../components/login/LoginScreen';
import DashboardRoutes from './DashboardRoutes';


const AppRouter = () => {

    const { user } = useContext(AuthContext);
    // console.log(user);

    return (
         <Router>
            <div>
                <Switch>

                    <PublicRoute
                        exact path="/login"
                        component={ LoginScreen }
                        isAuthenticated={ user.logged }
                    />

                    <PrivateRoute
                        path="/"
                        component={ DashboardRoutes }
                        isAuthenticated={ user.logged }
                    />

                    {/* <Route
                        exact path="/login"
                        component={ LoginScreen }
                    />  

                    <Route
                        path="/"
                        component={ DashboardRoutes }
                    />   */}
                   
                </Switch>
            </div>
        </Router>
    );
    
}

export default AppRouter
