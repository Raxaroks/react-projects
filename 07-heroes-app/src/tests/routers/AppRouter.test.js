import { mount } from "enzyme";
import { AuthContext } from "../../auth/AuthContext";
import AppRouter from "../../routers/AppRouter";

describe('Pruebas en el componente: AppRouter', () => {

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: false
        }
    };

    test('Debe de mostrarse el login si el usuario no se encuentra autenticado.', () => {
        const componentWrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <AppRouter/>
            </AuthContext.Provider>
        );

        expect(componentWrapper).toMatchSnapshot();
    });    

    test('Debe de mostrar la vista de Marvel si el usuario estÃ¡ autenticado.', () => {
        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged: true
            }
        };

        const componentWrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <AppRouter/>
            </AuthContext.Provider>
        );

        expect( componentWrapper.find('.navbar').exists() ).toBe( true );
    });    

});
