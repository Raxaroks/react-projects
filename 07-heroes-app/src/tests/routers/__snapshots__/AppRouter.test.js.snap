// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente: AppRouter Debe de mostrarse el login si el usuario no se encuentra autenticado. 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login
    </h1>
     
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
      style={
        Object {
          "borderRadius": 18,
        }
      }
    >
      Iniciar sesi√≥n
    </button>
  </div>
</div>
`;
